### PR TITLE
Add `use_local` flag to client.publish_flow()

### PR DESCRIPTION
**Pull request by Fanjia-Yan**
## Describe your changes and why you are making these changes
Change the `publish_flow()`, add a boolean parameter `use_local` such that when use_local is `True`, the user can publish workflow using local data. `use_local` is default to `False`. When it is `False`, the `publish_flow()` will error out.

## Related issue number (if any)
https://linear.app/aqueducthq/issue/ENG-2586/add-use-local-flag-to-clientpublish-flow

## Checklist before requesting a review
- [ ] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [ ] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [ ] If this is a new feature, I have added unit tests and integration tests.
- [ ] I have run the integration tests locally and they are passing.
- [ ] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

File: integration_tests/sdk/aqueduct_tests/param_test.py [hunk 1]
```diff
@@ -625,8 +625,15 @@ def append_row_to_df(df, row):
         output_artifact_list.append(output)
         input_data_list.append(input_df)
 
+    with pytest.raises(
+        InvalidUserActionException,
+        match="Cannot create a flow with local data."
+    ):
+        flow = publish_flow_test(
+            client, artifacts=output_artifact_list, name=flow_name(), engine=engine
```
Comment (kenxu95) [inline integration_tests/sdk/aqueduct_tests/param_test.py:633]: Let's use `client.publish_flow()` if we expect the flow to error before sending anything to the server.

File: sdk/aqueduct/client.py [hunk 2]
```diff
@@ -552,6 +556,9 @@ def publish_flow(
             source_flow:
                 Used to identify the source flow for this flow. This can be identified
                 via an object (Flow), name (str), or id (str or uuid).
+            use_local:
+                Used to identify if local data is used. In order to use `LocalData`, this has to be set
```
Comment (kenxu95) [inline sdk/aqueduct/client.py:560]: Maybe rephrase: "Must be set if any artifact in the flow is derived from local data."

File: sdk/aqueduct/artifacts/base_artifact.py [hunk 3]
```diff
@@ -14,6 +14,7 @@ class BaseArtifact(ABC):
     _dag: DAG
     _content: Any
     _from_flow_run: bool
+    _from_local_data: bool = False
```
Comment (kenxu95) [inline sdk/aqueduct/artifacts/base_artifact.py:17]: I would rather have this information on the underlying DAG's ArtifactMetadata object. See how this was done in https://github.com/aqueducthq/aqueduct/pull/1093/files, specifically `sdk/aqueduct/models/artifact.py`. 

The reason being that it's easier to think about all the construction fields are on the same data plane. `_from_flow_run` is different since the underlying DAG is fetched from the server. `_from_operator_type` is in the wrong place imo, so let's not follow that example.

File: sdk/aqueduct/client.py [hunk 4]
```diff
@@ -223,6 +223,9 @@ def create_param(self, name: str, default: Any, description: str = "") -> BaseAr
         check_explicit_param_name(self._dag, name)
         if isinstance(default, LocalData):
             default = extract_val_from_local_data(default)
+            artifact = create_param_artifact(self._dag, name, default, description)
+            artifact._from_local_data = True
```
Comment (kenxu95) [inline sdk/aqueduct/client.py:227]: Following the advice above, add a `from_local_data` argument to `create_param_artifact()`.

**Review by kenxu95 — CHANGES_REQUESTED** [review]
Lemme know if my comment isn't clear! Almost there though

**Comment by Fanjia-Yan** [comment]
Changes after reviews:

- Move `from_local_data` to ArtifactMetadata field that will be put into dag
- Only mark this field to True when the Artifact is a param artifact with local data input. 
- In `publish_flow`, after we construct the DAG subgraph, if there is any artifact in the subgraph that is a local data, raise Exception when `use_local` parameter is False which is default value.

File: sdk/aqueduct/models/artifact.py [hunk 5]
```diff
@@ -8,3 +8,4 @@ class ArtifactMetadata(BaseModel):
     id: uuid.UUID
     name: str
     type: ArtifactType
+    from_local_data: bool = False
```
Comment (kenxu95) [inline sdk/aqueduct/models/artifact.py:11]: You need to exclude this from the json config. See the reference PR I noted in my previous comment.

File: src/golang/lib/models/artifact.go [hunk 6]
```diff
@@ -20,10 +20,11 @@ const (
...
 type Artifact struct {
-	ID          uuid.UUID           `db:"id" json:"id"`
-	Name        string              `db:"name" json:"name"`
-	Description string              `db:"description" json:"description"`
-	Type        shared.ArtifactType `db:"type" json:"type"`
+	ID            uuid.UUID           `db:"id" json:"id"`
+	Name          string              `db:"name" json:"name"`
+	Description   string              `db:"description" json:"description"`
+	Type          shared.ArtifactType `db:"type" json:"type"`
+	FromLocalData bool                `db:"from_local_data" json:"from_local_data"`
```
Comment (kenxu95) [inline src/golang/lib/models/artifact.go:27]: oh we definitely shouldn't be propagating this field to the golang side. This is an SDK-only concept.

**Review by kenxu95 — CHANGES_REQUESTED** [review]

File: src/golang/lib/models/artifact.go (hunk 6 above)
Comment (Fanjia-Yan) [reply src/golang/lib/models/artifact.go]: Make sense that this is a SDK only object. Should be addressed!

File: sdk/aqueduct/client.py [hunk 7]
```diff
@@ -501,6 +508,7 @@ def publish_flow(
         k_latest_runs: Optional[int] = None,
         config: Optional[FlowConfig] = None,
         source_flow: Optional[Union[Flow, str, uuid.UUID]] = None,
+        use_local: Optional[bool] = False,
```
Comment (kenxu95) [inline sdk/aqueduct/client.py:511]: You might need to rebase. `list_params()` is gone now.

File: sdk/aqueduct/client.py [hunk 8]
```diff
@@ -552,6 +560,8 @@ def publish_flow(
             source_flow:
                 Used to identify the source flow for this flow. This can be identified
                 via an object (Flow), name (str), or id (str or uuid).
+            use_local:
+                Used to identify if local data is used. Must be set if any artifact in the flow is derived from local data.
```
Comment (kenxu95) [inline sdk/aqueduct/client.py:564]: nit: Let's scrape the first sentence. The second sentence is very clear already.

**Review by kenxu95 — APPROVED** [review]
Yay! Please address the nits before merging, and gl with the rebase let me know if you need help :)

File: sdk/aqueduct/client.py (hunk 7 above)
Comment (Fanjia-Yan) [reply sdk/aqueduct/client.py:511]: I will be doing the rebase in when `eng-2584` merge into `main` since I will be using from_local_data in that branch as well
